lib/HLSL/DxilPrecisePropagatePass.cpp

```diff
@@ -51,73 +54,169 @@ class DxilPrecisePropagatePass : public ModulePass {
 private:
   void PropagatePreciseOnFunctionUser(
       Function &F, DxilTypeSystem &typeSys,
-      std::unordered_set<Instruction *> &processedSet);
+      ValueSet &processedSet);
 };
 
 char DxilPrecisePropagatePass::ID = 0;
 
 }
 
 static void PropagatePreciseAttribute(Instruction *I, DxilTypeSystem &typeSys,
-    std::unordered_set<Instruction *> &processedSet);
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeOnPointer(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeOnPointerUsers(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeThroughGEPs(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ArrayRef<Value*> idxList, ValueSet &processedGEPs,
+    ValueSet &processedSet);
 
 static void PropagatePreciseAttributeOnOperand(
     Value *V, DxilTypeSystem &typeSys, LLVMContext &Context,
-    std::unordered_set<Instruction *> &processedSet) {
-  Instruction *I = dyn_cast<Instruction>(V);
-  // Skip none inst.
-  if (!I)
+    ValueSet &processedSet) {
+  // Skip values already marked.
+  if (!processedSet.insert(V).second)
     return;
 
-  FPMathOperator *FPMath = dyn_cast<FPMathOperator>(I);
-  // Skip none FPMath
-  if (!FPMath)
-    return;
+  if (V->getType()->isPointerTy()) {
+    PropagatePreciseAttributeOnPointer(V, typeSys, Context, processedSet);
+  }
 
-  // Skip inst already marked.
-  if (processedSet.count(I) > 0)
+  Instruction *I = dyn_cast<Instruction>(V);
```

Comment:
Should this be User* instead of Instruction*?

Reply:
For now, I limited the handling of non-instructions to pointer because I can't think of a valid non-pointer constant case where we need to propagate to arguments.